### ISSUE TITLE
AbstractForm backing-bean modified when ModalPopup closed.

### ISSUE THREAD

**Issue by rubberony** [viritin/viritin#186]
HI fives on the Viritin Addon, Matti et al.

I have a Grid of entities, the user selects one.  I then setEntity on AbstractForm and then openInModalPopup().  The form is displayed in its default form, with one Save button visible, fields to edit, and also importantly a close and maximise button on the modal window.

Change some field values and then close the dialog.  All appears fine, no changes to the Grid visible, but... Click on the same entity to edit again and Voila the changes I made in the first edit are back.

Digging down, this appears to be a side-effect of eager-validation making all fields immediate, that makes sense, and I really like the validation behaviour.

I've now overridden some methods in my Form, cited below.  I guess any implementation is quite unique per developer, and therefore not included, but I feel like I should have known about this via sample code or documentation, possibly even some sub-classes in the viritin addon?  Although I admit I haven't managed to go through all of the example projects yet.  Is there a recommended 'pattern' for this I've missed?

```java

    public UserForm() {
		super();
		// setEagerValidation(false);
		getDeleteButton().setVisible(true);
		getResetButton().setVisible(true);
		setResetHandler(this);
	}
    public MBeanFieldGroup<User> setEntity(User entity) {
		this.clone = entity.clone();
		fieldGroup = super.setEntity(entity);
		// fieldGroup.setBuffered(true); // doesn't work
		return fieldGroup;
	}

	@Override
	public Window openInModalPopup() {
		Window window = super.openInModalPopup();
		window.addCloseListener(this);
		return window;
	}

	@Override
	public void windowClose(CloseEvent e) {
		reset(null);
	}

	@Override
	public void onReset(User entity) {

		BeanItem<User> cloneBean = new BeanItem<User>(clone); // Construct a
																// bean Item
																// from the
																// cloned
																// instance we
																// saved!
		BeanItem<User> entityBean = new BeanItem<User>(entity);

		for (Object propertyId : cloneBean.getItemPropertyIds()) {
			Property itemProperty = cloneBean.getItemProperty(propertyId);
			Object defaultValue = itemProperty.getValue();

			Property entityProperty = entityBean.getItemProperty(propertyId);
			Object currentValue = entityProperty.getValue();

			if (!Objects.equal(defaultValue, currentValue)) {
				if (!entityProperty.isReadOnly()) {
					entityProperty.setValue(defaultValue);
				}
			}
		}
	}
```

**Comment by mstahv** [comment]
Hi,

Viritin forms always do non-buffered binding (because buffered binding is generally a bad idea). 

I typically do the cloning visa versa, edit the clone instead of the original bean. Or, if using JPA, I just [fetch a fresh instance from the DB](https://github.com/mstahv/jpa-addressbook/blob/master/src/main/java/org/example/MainView.java#L75-L94).

Also one pretty good option (except that you might cause one extra DB query, which shouldn't be a problem) is just to refresh the listing on reset as well.

BTW. Instead of using BeanItem to copy properties from bean to another, I'd definitely use [commons-beanutils](https://commons.apache.org/proper/commons-beanutils/).

cheers,
matti

**Comment by viydaag** [comment]
In other words, the discard() method from the MBeanFieldGroup will never work because the entity is always the latest version?

**Comment by mstahv** [comment]
Yep, discard() in Vaadin is related to buffering, which is not supported by Viritin.

If you want/need to support discard/cancel feature, you have still quite a lot of options. Here are couple which I have used.

 * Before passing the entity for editing, fetch a fresh one from your backend. If you have actively changing entities, this is also good for users, as there will less likely be concurrency issues (e.g. with JPA and optimistic locking). Also if you have JPA and relations, you probably have to do that to properly load lazy loaded references to other entities. Then if you click cancel/discard button in UI, you don't need to do anything. See [example here](https://github.com/mstahv/jpa-addressbook/blob/master/src/main/java/org/example/MainView.java#L75-L94).
 * A derived solution from the above is to clone the edited object in Java. If your object does not implement Cloneable, you can probably still use e.g. [copyProperties from Apache BeanUtils](https://commons.apache.org/proper/commons-beanutils/apidocs/org/apache/commons/beanutils/BeanUtils.html#copyProperties(java.lang.Object,%20java.lang.Object)). This is a universal way to create a "buffer" at bean level.
 * One way to workaround this issue is to do it "lazily". If your user clicks "discard/cancel" button, just refresh the data in components that used it. You shouldn't be afraid of performance impact, if this one extra query to backend chokes your application, you have problems anyways ;-)

**Comment by vonloxley** [comment]
Shouldn’t `discard()` be overriden and throw an  `UnsupportedOperationException`, then?

**Comment by mstahv** [comment]
That could indeed be a good option. And link to this thread for explanation.

**Comment by vonloxley** [comment]
Could you have a look at 2a1dc0a53339b2d5e0e89c376db0c9affac3c9a1, please? If this seems allright with you, I`ll open a pull-request.

**Comment by mstahv** [comment]
Looks good to me!

**Comment by rubberony** [comment]
Thanks guys I'm sure this change will help others starting out with Viritin.

**Comment by roeltje25** [comment]
This change is actually giving problems in my application. I am having the following setup. We have a model object with several values which is bound to a view using a MBeanFieldGroup which, of course, is unbuffered. However, when the data in the model object changes because of other reasons than sheer editing via the editing field. (e.g. when recalculating some parameters on a button click). I would like the beanfieldgroup to "refresh" the fields, as the underlying data has changed. We were using the .discard() method for that purpose which was working splendidly.... until it got forcefully deprecated and broken with an UnsupportedOperationException....

How could this be fixed?
1) perhaps reenable discard and adjust the javadoc?
2) perhaps introduce a refresh method that would force the fieldgroup to reload the underlying data
3) perhaps you should yell at me for doing this in the first place, although I am pretty confident about our framework.

**Comment by mstahv** [comment]
One option indeed could be to just go with JavaDoc warnings.

Could you share some kind of reduced usage example of what you are doing. That could give more insights to do the right fix.

**Comment by mstahv** [comment]
Now calling the unsupported methods just logs a warning message, but I'd still be eager to see your use case so I could come up with a real solution for you.

**Comment by roeltje25** [comment]
Ok. Sure. Our code does use a lot of custom abstract classes so the attached example won't run for anyone else. It contains a much simplified widget in our application that resides on a dashboard. The view contains a design file with a Date field bound to the field value thing in the model. The model is bound to the design using a subclass of MBeanFieldGroup. At viewenter (see presenter) a possibly long business logic process is started that will yield a new version of the date field in the model. a modelChanged event will trigger a method in the view that will refresh the design by calling discard on the fieldgroup. Now of course our widgets tend to have much more fields to be refreshed. For just a date field the whole setup may seem a bit overdeveloped ;-)
[simple.zip](https://github.com/viritin/viritin/files/439375/simple.zip)
